### ISSUE TITLE
Ensure that `waitForResponse` and similar methods are awaited after triggering condition

### ISSUE THREAD

**Issue by unlikelyzero** [mskelton/eslint-plugin-playwright#199]
In order to make tests less flaky, any call to page.waitForResponse should be wrapped in a promise.all of the action which invoked the network response.

More here: https://github.com/microsoft/playwright/issues/5470#issuecomment-1285640689

**Comment by mxschmitt** [comment]
Just some thoughts:

```ts
// Good:
const [response] = await Promise.all([
  page.locator("button").click(),
  page.waitForResponse("https://example.com/api/search")  
]);

// Good
const mySearchPromise = page.waitForResponse("https://example.com/api/search")
await page.locator("button").click()
await mySearchPromise;

// Bad

await page.waitForResponse("https://example.com/api/search")
```

**Comment by thomas-phillips-nz** [comment]
Yeah we use the `create promise -> perform action -> await promise` pattern a lot in our codebase.

**Comment by mskelton** [comment]
@unlikelyzero Interestingly enough, the Playwright docs don't even use `Promise.all`

```ts
const responsePromise = page.waitForResponse('https://example.com/resource');
await page.getByText('trigger response').click();
const response = await responsePromise;
```

_@mxschmitt While outside the scope of this issue, seems like a good change to make to the docs to use `Promise.all` in the Playwright docs._

My main concern around this rule is doing it in a way that doesn't produce false positives/negatives.

```ts
const responsePromise = page.waitForResponse('https://example.com/resource');
await doStuffThatFiresResponse() // Is this valid or invalid?
await responsePromise

const responsePromise = page.waitForResponse('https://example.com/resource');
await myPageObject.doStuff() // Is this valid or invalid?
await responsePromise

await Promise.all([
  page.waitForResponse('https://example.com/resource'),
  doStuff() // How do we know that do stuff actually is firing the request?
])
```

I suppose the way this could be implemented would be to ensure that either:

1. The `waitForResponse` (and other similar methods) are in a `Promise.all`
2. The `waitForReponse` is not awaited until **after** another awaited method, regardless of what the other method does.

If we did that, we might not catch everything that is actually an error, for example:

```ts
const responsePromise = page.waitForResponse('https://example.com/resource');
const isVisible = await page.locator('.foo').isVisible()
await responsePromise
await page.locator('.foo').click() // This fires the request
```

Would not be marked as an error, even though it is invalid. That's a super contrived example and unlikely to happen in the real world, so the value of catching the common mistakes seems like it could be implemented in a way that catches 95% of all mistakes.

**Comment by unlikelyzero** [comment]
> Would not be marked as an error, even though it is invalid. That's a super contrived example and unlikely to happen in the real world, so the value of catching the common mistakes seems like it could be implemented in a way that catches 95% of all mistakes.

This seems like a reasonable tradeoff

**Comment by tphillips-ict** [comment]
This almost seems like it's outside the realm of static code analysis, promises can be created and awaited in completely different contexts. This is kind of bordering on runtime analysis.

**Comment by mxschmitt** [comment]
This was an intentional change, see here for our reasoning: https://github.com/microsoft/playwright/pull/19154

**Comment by mskelton** [comment]
Only took two years, but this is finally fixed 😔
